### ISSUE TITLE
UNAUTHENTICATED response for admin.firestore.collection.doc.get running in parallel 

### ISSUE THREAD

**Issue by nilushan-bluekey** [googleapis/nodejs-firestore#1738]
- What you're trying to do

IoTCore devices send events and requests through pubsub. A GKE workload with multiple pods subscribes to the messages and responds to the IoTCore devices. 
**Workload Runtimes**

> - node 12.18 
> - firebase-admin: 8.12.1

When a surge of IoTCore device messages occurs, each pod receives a larger number of messages to handle. This causes a few requests in parallel to firestore via the nodejs-firestore sdk.  During this load, we are receiving a few errors.

In addition to firestore failing, pubsub message ack fails to go through as well. 
The code works fine when there is no surge in load. 

- What code you've already tried

Mainly get and update firestore collections.  This call is run in parallel multiple times. 

```
  
public get(collection: string, doc: string | undefined): Promise<any> {

    return new Promise((resolve, reject) => {
      if (doc) {
        this.firestoreClient
          .collection(collection)
          .doc(doc)
          .get()
          .then(snapshot => {
            const retItem = snapshot.data();
            if (retItem) {
              resolve(retItem);
            } else {
              reject(new Error('entity not found'));
            }
          })
          .catch(err => {
            log.error('Error getting documents ' + err);
            reject(err);
          });
      } else {
        reject(new Error('document id undefined'));
      }
    });
  }
```
- Any error messages you're getting

```
16 UNAUTHENTICATED: Failed to retrieve auth metadata with error: request to https://www.googleapis.com/oauth2/v4/token failed, reason: Client network socket disconnected before secure TLS connection was established
```

Errors not related to firestore
```
4 DEADLINE_EXCEEDED: Deadline exceeded
    at AckQueue._sendBatch (/usr/src/app/node_modules/@google-cloud/pubsub/build/src/message-queues.js:178:19)
    at runMicrotasks (<anonymous>)
    at runNextTicks (internal/process/task_queues.js:62:5)
    at processImmediate (internal/timers.js:429:9)
```

- Question

Are there any issues in sending a few firestore get collection calls in parallel? Should applications be developed to minimize calls in parrellel?

**Comment by dconeybe** [comment]
Hi @nilushan-bluekey. I don't have any immediate ideas but I'll take a look into it and reply back.

**Comment by nilushan-bluekey** [comment]
Thank you very much @dconeybe 

I will do an audit of the code as you suggested. There is a good chance that some promises do not await properly.

**Comment by dconeybe** [comment]
Since it's been a week I'm going to close this issue; however, feel free to comment if you are unsatisfied with this resolution, or file a new ticket if the issue resurfaces. Thanks!
